### PR TITLE
Panel collapse fixes

### PR DESCRIPTION
**Pull request by amcclain**
Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [ ] Added CHANGELOG entry, or determined not required.
- [ ] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [x] Updated doc comments / prop-types, or determined not required.
- [ ] Reviewed and tested on Mobile, or determined not required.
- [ ] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

**Comment by amcclain** [comment]
Ended up pushing further on this than intended - was just about to merge our earlier PR in #3478 but then saw the bug Saba reported on that PR around opening collapsed panels in a modal.  That + earlier convo about modalToggle button led to a more substantial refactor, which should be looked over carefully!

* Opening a new PR in case we run into issues here and want to go back to the earlier one and file other issues as tickets for later (but hoping we can merge this one)
* I don't think that anything here is relevant to mobile

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,8 @@
 ### 🎁 New Features
 
 * Added new `Panel.headerClassName` prop for straightforward CSS manipulation of panel's header.
+* `Panel.collapsedTitle` will now display the header even if there is no title when expanded.
+* Added new `Panel.collapsedIcon` prop to display the header even if there is no icon when expanded.
```
Comment (amcclain) [inline CHANGELOG.md:9]: I am happy to update CL once we settle on a final scope here

File: desktop/cmp/panel/Panel.ts [hunk 2]
```diff
@@ -173,15 +177,15 @@ export const [Panel, panel] = hoistCmp.withFactory<PanelProps>({
...
+                style: {display: isRenderedCollapsed ? 'none' : 'flex'},
                 items: Children.toArray([
                     parseToolbar(tbar),
                     ...castArray(children),
                     parseToolbar(bbar)
                 ])
             });
         }
-        if (!collapsed) wasDisplayed.current = true;
+        if (!isRenderedCollapsed) wasDisplayed.current = true;
```
Comment (amcclain) [inline desktop/cmp/panel/Panel.ts:188]: I want to look a bit more closely at this + interaction with refreshContext - I think that this change is an improvement there, in that modal panels will get refreshed - but didn't fully trace out that part of panelModel

File: desktop/cmp/panel/PanelModel.ts [hunk 3]
```diff
@@ -157,8 +160,13 @@ export class PanelModel extends HoistModel {
         return !!this.modalSupportModel;
     }
 
+    /** True when both collapsed and not currently in a modal - i.e. *really* collapsed. */
+    get isRenderedCollapsed(): boolean {
+        return this.collapsed && !this.isModal;
+    }
```
Comment (amcclain) [inline desktop/cmp/panel/PanelModel.ts:166]: Struggled with this name a bit (and the newly qualified comment on `collapsed` above)...  I'm OK with where it landed but happy to discuss any alternatives.

File: desktop/cmp/panel/PanelModel.ts [hunk 4]
```diff
@@ -340,6 +348,10 @@ export class PanelModel extends HoistModel {
         return this.side === 'top' || this.side === 'left';
     }
 
+    get isCollapsedToLeftOrRight(): boolean {
+        return this.isRenderedCollapsed && !this.vertical;
+    }
```
Comment (amcclain) [inline desktop/cmp/panel/PanelModel.ts:353]: I know this is wordy - but I got the sense of `vertical` wrong on a few different occasions and wanted something that was clear enough for me to keep straight 😉

File: desktop/cmp/panel/PanelModel.ts [hunk 5]
```diff
@@ -227,8 +235,8 @@ export class PanelModel extends HoistModel {
         this.refreshMode = refreshMode;
         this.showSplitter = showSplitter;
         this.showSplitterCollapseButton = showSplitterCollapseButton;
-        this.showHeaderCollapseButton = showHeaderCollapseButton;
-        this.showModalToggleButton = showModalToggleButton;
+        this.showHeaderCollapseButton = collapsible && showHeaderCollapseButton;
+        this.showModalToggleButton = modalSupport && showModalToggleButton;
```
Comment (amcclain) [inline desktop/cmp/panel/PanelModel.ts:239]: I think it's an improvement to force these to false when never gonna be supported - cuts down on checking code elsewhere.  Unfortunately there's some asymmetry in that showModalToggle will always be correct, but showHeaderCollapseButton must still be combined with `!isModal` within `PanelHeader`.

Could have saved the supplied config to a _showHeaderCollapse and added a getter, wasn't sure it was worth it.

File: desktop/cmp/panel/impl/PanelHeader.ts [hunk 6]
```diff
@@ -41,57 +62,57 @@ export const panelHeader = hoistCmp.factory({
...
-                    hbox({
-                        className: 'xh-panel-header__items',
-                        items: [
-                            ...(!collapsed || isModal ? headerItems : []),
-                            modalButton({panelModel}),
-                            collapseButton({panelModel})
-                        ],
-                        onDoubleClick: e => e.stopPropagation()
-                    })
+                    ...displayedHeaderItems
```
Comment (amcclain) [inline desktop/cmp/panel/impl/PanelHeader.ts:84]: In the vertical case we're not wrapping with ` className: 'xh-panel-header__items'` like we are below.

We could add for consistency, but it looked like we'd then need to add more specific styles to effectively undo the effect of the styles applied by the class, which are targeted at the horizontal orientation.  Wasn't sure that the inconsistency would throw off anything else, and in this case the only two header items that are possible are the two toggles.

File: desktop/cmp/panel/PanelModel.ts (hunk 3 above)
Comment (TomTirapani) [reply desktop/cmp/panel/PanelModel.ts:166]: This distinction and naming makes sense to me 👍

File: desktop/cmp/panel/impl/PanelHeader.ts [hunk 7]
```diff
@@ -41,57 +62,57 @@ export const panelHeader = hoistCmp.factory({
...
+                displayedIcon,
                 displayedTitle
                     ? box({
                           className: titleCls,
+                          flex: 1,
                           item: span({className: `${titleCls}__inner`, item: displayedTitle})
                       })
-                    : null
+                    : filler(),
+                hbox({
```
Comment (TomTirapani) [inline desktop/cmp/panel/impl/PanelHeader.ts:103]: Curious about the decision to move these to the bottom for L/R panel? At the top the collapse button is at the same Y position as when expanded, but that paradigm makes less sense now the modal button is involved.

Is it just a desire to not show up to 3 icons before the title? Or wanting the flow to remain Icon > Title > Space > Buttons? I don't feel too strongly either way.

File: desktop/cmp/panel/impl/PanelHeader.ts [hunk 8]
```diff
@@ -41,57 +62,57 @@ export const panelHeader = hoistCmp.factory({
...
     }
 });
 
-const collapseButton = hoistCmp.factory(({panelModel}) => {
-    const {showHeaderCollapseButton, collapsible, isModal} = panelModel as PanelModel;
-    if (!showHeaderCollapseButton || !collapsible || isModal) return null;
-
-    const {vertical, collapsed, contentFirst} = panelModel,
+const collapseToggleButton = hoistCmp.factory(({panelModel}) => {
+    const {vertical, collapsed, contentFirst} = panelModel as PanelModel,
```
Comment (TomTirapani) [inline desktop/cmp/panel/impl/PanelHeader.ts:115]: Great that we could move this visibility logic out of these internal components (and completely remove the wrapper below) :)

File: desktop/cmp/panel/PanelModel.ts (hunk 4 above)
Comment (TomTirapani) [reply desktop/cmp/panel/PanelModel.ts:353]: If we want a shorter alternative, maybe `isLeftRightCollapsed`? Not sure that's necessarily any better :)

**Comment by TomTirapani** [comment]
Looks great to me 👍

**Comment by TomTirapani** [comment]
On a semi-related aside, have we considered rendering the space left behind when a panel is modal in a grey / bg-alt / tiled pattern? Just to make it look more obviously "empty". I think the plain background doesn't adequately convey that something has been "removed".

image

Just a thought, not necessarily something we need to address here.

File: desktop/cmp/panel/impl/PanelHeader.ts (hunk 6 above)
Comment (lbwexler) [reply desktop/cmp/panel/impl/PanelHeader.ts:84]: just a thought -- would it be helpful to just have a seperate component for the vertical collapsed case?

File: desktop/cmp/panel/PanelModel.ts (hunk 3 above)
Comment (saba-mo) [reply desktop/cmp/panel/PanelModel.ts:175]: Can we add (not modal) after "home" in the description of collapsed? Might help devs get the idea faster.

File: desktop/cmp/panel/PanelModel.ts (hunk 4 above)
Comment (saba-mo) [reply desktop/cmp/panel/PanelModel.ts:370]: I like Tom's suggestion of isLeftRightCollapsed. And I agree "vertical" is confusing - it makes me think of the LR collapsed header, which runs vertically.

**Review by saba-mo — APPROVED** [review]

**Pull request merged**